Boron removal model

aladshaw3 opened the pull request:
## Fixes/Addresses:

#492 

## Summary/Motivation:

Creates a simplified equilibrium model for boron removal. Works with or without protons and hydroxide as explicit species in the property package. Calculates pH change and boron/borate distribution based on addition of a caustic additive defined by the user. Works with the ion_prop_pack in watertap and technically will work with the generic property package in IDAES. 

## Changes proposed in this PR:
- Adds boron removal and associated tests
- Modifies the bounds on the ion_prop_pack (REQUIRED to facilitate the pH dependence) 
- Updates ion_prop_pack tests 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

Commit:
Started with basic boron removal instead of complex model

Commit:
Added some error checking for config

Commit:
Created variables for model. Need to refine and add constraints

aladshaw3 committed:
Merge branch 'watertap-org:main' into boron-removal-zo

Commit:
Saving work

Commit:
Had to change lower bounds on ion prop pack to get convergence.

Commit:
Almost done building constraints, but ran into an error with generic prop pack.

Commit:
Updated ion prop pack tests due to changes in lower bounds

Commit:
Finialized the constraints. Need to add scaling and tests

Commit:
Updated config info

Commit:
Saving work

aladshaw3 committed:
Merge branch 'watertap-org:main' into boron-removal-zo

Commit:
Added temperature relation for reactions.

Commit:
Created propagate state function to initialize some of the difficult to scale properties.

Commit:
Need to finish scaling and testing

Commit:
Saving work

Commit:
Finished scaling and testing. Seems ready to go.

Commit:
Started test files.

Commit:
Finalized initial testing

Commit:
created all tests

Commit:
Ran black on all

Commit:
Trying to fix doc build error

Commit:
Trying to fix doc build error

Commit:
Reran black

lbibl commented on watertap/unit_models/boron_removal.py:
```diff
@@ -0,0 +1,1123 @@
+
+        # Next, get the base units of measurement from the property definition
+        units_meta = self.config.property_package.get_metadata().get_derived_units
+
```
Can this units_meta be removed since all units are/can be adopted from `pyunits` directly?

andrewlee94 replied:
This functionality is intended to allow for general models where the units of measurement are not know a priori - i.e. in many cases the units of measurement are determined by the property package, and thus the units definition in the unit model will need to change depending on what the property package is using.

lbibl commented on watertap/unit_models/boron_removal.py:
```diff
@@ -0,0 +1,1123 @@
+        )
+        def eq_electroneutrality(self, t):
+            Alk = 0
+            for j in self.ion_charge:
```
Isn't `self.ion_charge` the charge values (e.g., +1 or -1) instead of the ion names?  but it must be right as the code passes properly.

aladshaw3 replied:
ion_charge is a pyomo Param indexed by name, so the values of j will be the names. 

If it were just a list, then you would be correct.

lbibl commented on watertap/unit_models/boron_removal.py:
```diff
@@ -0,0 +1,1123 @@
+                    self.control_volume.properties_out[t].flow_vol_phase[p] * boron_out
+                )
+
+                loss_rate = input_rate - exit_rate
```
Just a question-  This constraint seems to be only for a unit conversion essentially ?

aladshaw3 replied:
No, this constraint is basing the loss rate off of the solution to the boron concentration inside the unit at the given pH and applies that loss rate such that we achieve that end result.

aladshaw3 replied:
It does include in it a unit conversion, but that is just to ensure that the units always work out regardless of how the property package was setup.

lbibl reviewed (dismissed):
I made a few minor comments. The overall content makes sense looks good to me.  A general comment: this seems to calculate Boron speciation upon a pH increase and I assume this unit is usually adapted to a RO to make the RO reaction of Boron more efficient; if this is correct, should we name it something else rather than "Boron removal"?  (Correct me if I am wrong. )    I approve in the scope of this pH increase treatment scheme.

adam-a-a commented on watertap/unit_models/boron_removal.py:
```diff
@@ -0,0 +1,1123 @@
+                )
+
+                loss_rate = input_rate - exit_rate
+                return self.control_volume.mass_transfer_term[t, p, j] == -loss_rate
```
I skimmed this quickly so may be wrong, but can these be condensed? It looks like you might only need the conditional to decide which component molarity will be converted.

e.g.,: 
```
if j == self.proton_name_id:
    molarity = self.mol_H[t]
elif j == self.hydroxide_name_id:
   molarity = self.mol_OH[t]
.....
```

adam-a-a replied:
Wouldn't change functionality but would reduce number of lines

aladshaw3 replied:
Yes, I think that may work for most of them.

adam-a-a commented on watertap/unit_models/boron_removal.py:
```diff
@@ -0,0 +1,1123 @@
+            units=pyunits.mol / pyunits.m**3,
+            doc="Resulting molarity of Borate",
+        )
+
```
So, aren't these the molarities at the outlet? Is there a reason why these couldn't be represented by `self.properties_out[t].conc_mol_phase_comp`?

aladshaw3 replied:
My reasoning was initially for testing purposes to make sure that the internal speciation problem was solved correctly without explicitly needing to couple with the properties. After I was satisfied with the results, I made constraints on mass transfer terms to essentially enforce this equality. 

Doing it in the way I am doing it also makes the unit model a little less dependent on the scaling of the property package, as I can scale the internal representations separately from the scaling employed by the package. I also get better control of the units on the variables, which can be helpful if this is used with multiple property packages.

adam-a-a replied:
I'm all about approaches that make things more helpful although this might deviate from our conventions. I don't feel strongly about changing what you have necessarily, but we probably want @andrewlee94 and @TimBartholomew 's opinions on this too.

aladshaw3 replied:
Will update names to standard naming conventions to make this more clear.

adam-a-a commented on watertap/unit_models/boron_removal.py:
```diff
@@ -0,0 +1,1123 @@
+                            ]
+                            * self.control_volume.properties_in[t0].mw_comp[ind[1]]
+                        )
+
```
shouldn't this already be handled by the prop model?

adam-a-a replied:
Oh, I guess you are updating these values to match the instance relevant to the unit model.

aladshaw3 replied:
Depending on which property package is used, these may be initialized differently (or at different times). Thus, I have this function to reinitialize them to the most relevant values for the model. It helps **significantly** with scaling. I highly recommend doing this reinitialization in any unit model for setting initial property values relevant to the unit model's problem.

adam-a-a replied:
I think I like it. =)

Commit:
Merge branch 'main' into boron-removal-zo

Commit:
Merge branch 'main' into boron-removal-zo

Commit:
Merge branch 'main' into boron-removal-zo

Commit:
Updated per discussions

aladshaw3 commented:
@adam-a-a @TimBartholomew @lbibl 

Updated the unit model per suggestions during our meeting. Also changed one of the unit tests to solve for the dosage of caustic additive needed to achieve a specific outlet rate of boron.

Commit:
Added some notes in the solve

Commit:
Ran black... again...

aladshaw3 commented on watertap/unit_models/boron_removal.py:
```diff
@@ -0,0 +1,1141 @@
+                return self.control_volume.mass_transfer_term[t, p, j] == -loss_rate
+            elif j == self.cation_name_id:
+                c_out = pyunits.convert(
+                    self.caustic_cation_charge * self.caustic_dose[t] / self.caustic_mw,
```
```suggestion
                    self.caustic_dose[t] / self.caustic_mw,

```

aladshaw3 committed:
Update watertap/unit_models/boron_removal.py

typo in equation

Commit:
Ran black once again...

aladshaw3 commented:
**NOTE**: This PR will be paused pending some revisions and updates per conversations with the Chem Team. 

- [x] Update model to enforce user provide a caustic cation species in the state variables (so we can properly track changes in salinity from inlet to outlet)
- [x] Update chemical dose to a total mass of additive (rather than concentration)
- [x] Update unit to have a volume variable set by the user (used to determine retention time)

@lbibl Let me know if I am missing anything here we discussed.

aladshaw3 committed:
Merge branch 'watertap-org:main' into boron-removal-zo

Commit:
Updated model to force user to provide the cation associated with the chemical additive to the state vars

Commit:
Fixed all model issues. Need to add final vars and run black.

Commit:
Updated with new vars and ran black

Commit:
Merge branch 'main' into boron-removal-zo

Commit:
Added unit conversion in case it is necessary

lbibl reviewed (dismissed):
Looking good to me.

Commit:
Merge branch 'main' into boron-removal-zo

bknueven commented on watertap/unit_models/boron_removal.py:
```diff
@@ -0,0 +1,1202 @@
+                )
+
+                loss_rate = input_rate - exit_rate
+                return self.control_volume.mass_transfer_term[t, p, j] == -loss_rate
```
I'm not a huge fan of the repeated code in this constraint formulation. It could be eliminated by using a simple dictionary mapping:
```python
{ self.hydroxide_name_id : self.conc_molOH,
...
}
```

aladshaw3 replied:
I will see about condensing this down.

bknueven commented on watertap/unit_models/boron_removal.py:
```diff
@@ -0,0 +1,1202 @@
+            )
+            iscale.set_scaling_factor(
+                blk.conc_mol_Borate[t], 100 / blk.conc_mol_Borate[t].value
+            )
```
It seems a bit dangerous to rescale these purely based on value without an upper limit.

aladshaw3 replied:
It should not really be much of an issue for these variables. Most of these 'values' will not even approach 1. I guess I could add an upper limit to the variables to be safe.

bknueven replied:
Sorry, I mean an upper limit on `100 / var.value`, i.e., the applied scaling factor.

bknueven commented on watertap/unit_models/boron_removal.py:
```diff
@@ -0,0 +1,1202 @@
+                                * self.control_volume.properties_in[t0].mw_comp[j]
+                                for j in self.config.property_package.component_list
+                            )
+                        )
```
It might make sense to move this to a helper function so reduce the indentation two levels, making the code easier to read and reason about (e.g., am I still in this `if` block?)

aladshaw3 replied:
Sure, I can do that.

bknueven commented on watertap/unit_models/boron_removal.py:
```diff
@@ -0,0 +1,1202 @@
+    def propogate_initial_state(self):
+        units_meta = self.config.property_package.get_metadata().get_derived_units
+        i = 0
+        for t in self.control_volume.properties_in:
```
Since it seems like we need `i`, it would be better to use `enumerate` here, e.g.:
```suggestion
        for i, t in enumerate(self.control_volume.properties_in):
```

That said, since it seems like you just use this to grab the first time index, you might instead just put this before the loop:
`t0 = self.flowsheet().time.first()`

bknueven reviewed (commented):
I can't really speak to the model, but I have a few comments on the code and a comment on the auto-scaling the initialization is doing.

Commit:
Merge branch 'main' into boron-removal-zo

Commit:
Resolving issues from comments

Commit:
Merge branch 'main' into boron-removal-zo
